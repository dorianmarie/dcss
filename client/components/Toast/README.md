This component is derived from https://github.com/academia-de-codigo/react-semantic-toasts. That component served well until it didn't.

