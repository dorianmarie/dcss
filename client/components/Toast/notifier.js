import Store from './store';

const store = new Store();
let id = 0;

function notifier(item, onClose, onClick, onDismiss) {
  id += 1;
  store.add({ id, onClose, onClick, onDismiss, ...item });
}

export { notifier, store };
