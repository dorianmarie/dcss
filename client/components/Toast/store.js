class Store {
  items = [];
  subscribers = [];

  subscribe(cb) {
    this.subscribers.push(cb);
  }

  unsubscribe(cb) {
    this.subscribers = this.subscribers.filter(subscriber =>
      subscriber !== cb ? subscriber : undefined
    );
  }

  notify() {
    this.subscribers.forEach(subscriber => subscriber(this.items));
  }

  add(item) {
    this.items.push(item);
    this.notify();
  }

  clear() {
    this.items.length = 0;
    this.notify();
  }

  remove(item) {
    this.items = this.items.filter(storeItem =>
      storeItem !== item ? storeItem : undefined
    );
    this.notify();
  }

  get data() {
    return this.items;
  }
}

export default Store;
