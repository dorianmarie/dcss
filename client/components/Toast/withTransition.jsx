import React from 'react';
import PropTypes from 'prop-types';
import { Transition } from '@components/UI';

const OPEN_TIME = 500;
const CLOSE_TIME = 1000;

export default function withTransitions(Component) {
  class ToastTransition extends React.Component {
    static propTypes = {
      id: PropTypes.number,
      onClose: PropTypes.func,
      openAnimation: PropTypes.string,
      closeAnimation: PropTypes.string,
      time: PropTypes.number
    };

    static defaultProps = {
      time: 2000
    };

    state = {
      visible: false,
      time: OPEN_TIME,
      animation: this.props.openAnimation
    };

    timeout = null;

    onClose = () => {
      this.setState(
        prevState => ({
          visible: !prevState.visible,
          animation: this.props.closeAnimation,
          time: CLOSE_TIME
        }),
        () => {
          setTimeout(() => {
            if (this.timeout) {
              clearTimeout(this.timeout);
            }

            this.props.onClose(this.props.id);
          }, CLOSE_TIME);
        }
      );
    };

    componentDidMount() {
      // schedule auto closing of toast
      if (this.props.time) {
        this.timeout = setTimeout(this.onClose, this.props.time);
      }

      // start animation as soon as toast is mounted in the dom
      this.setState({ visible: true });
    }

    render() {
      const {
        id,
        openAnimation,
        closeAnimation,
        time: timeProp,
        onClose,
        ...props
      } = this.props;
      const { time, visible, animation } = this.state;
      const styles = {
        marginBottom: '1em'
      };

      return (
        <Transition animation={animation} duration={time} visible={visible}>
          <div style={styles} role="presentation">
            <Component {...props} onClose={this.onClose} />
          </div>
        </Transition>
      );
    }
  }

  return ToastTransition;
}
