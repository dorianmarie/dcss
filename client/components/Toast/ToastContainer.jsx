import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Toast from './Toast';
import { store } from './notifier';
import Identity from '@utils/Identity';
import './Toast.css';

const closeAnimations = {
  'top-right': 'fly left',
  'top-center': 'fly down',
  'top-left': 'fly right',
  'bottom-right': 'fly left',
  'bottom-center': 'fly up',
  'bottom-left': 'fly right'
};

class ToastContainer extends Component {
  static propTypes = {
    animation: PropTypes.string,
    className: PropTypes.string,
    position: PropTypes.oneOf(Object.keys(closeAnimations))
  };

  static defaultProps = {
    animation: null,
    className: '',
    position: 'top-right'
  };

  state = {
    toasts: []
  };

  componentDidMount() {
    store.subscribe(this.updateToasts);
  }

  componentWillUnmount() {
    store.unsubscribe(this.updateToasts);
  }

  onClose = id => {
    const toast = this.state.toasts.find(value => value.id === id);

    if (!toast) {
      return;
    }

    store.remove(toast);

    if (toast.onClose) {
      toast.onClose();
    }
  };

  updateToasts = toasts => {
    this.setState({
      toasts
    });
  };

  render() {
    const { animation: containerAnimation, position, className } = this.props;
    const { toasts } = this.state;
    const { onClose } = this;
    return toasts.length ? (
      <div className={`ui-alerts ${position} ${className}`}>
        {toasts.map(toast => {
          const {
            animation,
            className = '',
            color,
            description = '',
            icon,
            id,
            list,
            onClick,
            onDismiss,
            size,
            time,
            title = '',
            type = ''
          } = toast;

          const key = Identity.key(toast);
          const openAnimation = animation || containerAnimation || 'pulse';
          const closeAnimation = closeAnimations[position];
          const toastProps = {
            className,
            closeAnimation,
            color,
            description,
            icon,
            id,
            key,
            list,
            onClick,
            onClose,
            onDismiss,
            openAnimation,
            size,
            time,
            title,
            type
          };

          return <Toast {...toastProps} />;
        })}
      </div>
    ) : null;
  }
}

export default ToastContainer;
