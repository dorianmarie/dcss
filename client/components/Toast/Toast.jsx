import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Header, Icon, Message } from '@components/UI';
import withTransition from './withTransition';

const icons = {
  info: 'announcement',
  success: 'checkmark',
  error: 'remove',
  warning: 'warning circle'
};

function Toast(props) {
  const {
    className = '',
    description,
    title,
    /* style = {}, */ ...rest
  } = props;

  const onDismiss = e => {
    e.stopPropagation();
    rest.onDismiss();
    rest.onClose();
  };

  const icon = rest.icon || icons[rest.type];
  const header = title || null;
  const content = (
    <Fragment>
      {header ? (
        <Header size="small">
          {icon ? <Icon className={icon} /> : null}
          <Header.Content>{header}</Header.Content>
        </Header>
      ) : null}
      {description || null}
    </Fragment>
  );

  const type = rest.type ? { [rest.type]: true } : {};
  const floating = true;
  const messageProps = {
    ...rest,
    ...type,
    onDismiss,
    content,
    floating
  };

  // This is in the header
  delete messageProps.icon;

  return (
    <div className={className}>
      <Message {...messageProps} />
    </div>
  );
}

Toast.propTypes = {
  className: PropTypes.string,
  description: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.string,
    PropTypes.node
  ]),
  icon: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  onDismiss: PropTypes.func,
  onClose: PropTypes.func,
  style: PropTypes.object,
  type: PropTypes.string,
  title: PropTypes.node
};

Toast.defaultProps = {
  onDismiss: () => undefined,
  onClose: () => undefined,
  icon: undefined
};

export default withTransition(Toast);
