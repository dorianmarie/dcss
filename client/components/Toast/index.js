import ToastContainer from './ToastContainer';
import { notifier, store } from './notifier';
notifier.store = store;
export { ToastContainer, notifier };
