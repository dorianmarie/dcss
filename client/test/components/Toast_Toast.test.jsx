import React from 'react';
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useLayoutEffect: jest.requireActual('react').useEffect
}));

import {
  fetchImplementation,
  mounter,
  reduxer,
  serialize,
  snapshotter,
  state
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';

import {
  act,
  fireEvent,
  prettyDOM,
  render,
  screen,
  waitFor
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Toast from '../../components/Toast/Toast.jsx';

const original = JSON.parse(JSON.stringify(state));
let container = null;
let commonProps = null;
let commonState = null;

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

afterAll(() => {
  jest.restoreAllMocks();
});

beforeEach(() => {
  jest.useFakeTimers();
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);

  commonProps = {};
  commonState = JSON.parse(JSON.stringify(original));
});

afterEach(() => {
  jest.runOnlyPendingTimers();
  jest.useRealTimers();
  jest.resetAllMocks();
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  commonProps = null;
  commonState = null;
});

test('Toast', () => {
  expect(Toast).toBeDefined();
});

test('Render 1 1', async done => {
  const Component = Toast;

  const props = {
    ...commonProps
  };

  const state = {
    ...commonState
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  const { asFragment } = render(<ConnectedRoutedComponent {...props} />);
  expect(asFragment()).toMatchSnapshot();

  done();
});
