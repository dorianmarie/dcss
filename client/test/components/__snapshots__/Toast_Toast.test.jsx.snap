// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1 1 1`] = `
<DocumentFragment>
  <div
    class="fade animating in visible transition"
    role="presentation"
    style="margin-bottom: 1em; animation-duration: 500ms;"
  >
    <div
      class=""
    >
      <div
        class="ui floating message"
      >
        <i
          aria-hidden="true"
          class="close icon"
        />
        <div
          class="content"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
